File: Cpp/001~100/032. Longest Valid Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int res = 0, cnt = 0;
        stack<int> S;
        for (auto &c : s){
            if (c == '(') {
                ++cnt;
                S.push(0); 
            }
            else if (S.empty() || !cnt) {
                cnt = 0;
                if( !S.empty() ) S.pop();
            }
            else{
                int cur = 2;
                while( !S.empty() && S.top() ){
                    cur += S.top();
                    S.pop();
                }
                S.pop();
                while( !S.empty() && S.top() ){
                    cur += S.top();
                    S.pop();
                }
                S.push(cur);
                cnt--;
                res = max(res, cur);
            } 
        }
        return res;
    }
};



//////////////////////solution 2/////////////////////////////////////


class Solution {
public:
    int longestValidParentheses(string s) {
        int res = 0, left = -1, n = s.length();
        stack<int> S;
        for (int i = 0; i < n; i++){
            if (s[i] == '(') S.push(i); 
            else if (S.empty()) left = i;
            else {
                S.pop();
                if (S.empty())  res = max(res, i-left);
                else res = max(res, i-S.top());
            }
        }
        return res;
    }
};

File: Go/032. Longest Valid Parentheses.go
func longestValidParentheses(s string) int {
	dp := make([]int, len(s))
	res := 0
	for i := 1; i < len(s); i++ {
		if s[i] == ')' {
			if s[i-1] == '(' {
				if i-2 >= 0 {
					dp[i] = 2 + dp[i-2]
				} else {
					dp[i] = 2
				}
			} else if i-1-dp[i-1] >= 0 && s[i-1-dp[i-1]] == '(' {
				if i-1-dp[i-1]-1 >= 0 {
					dp[i] = dp[i-1] + 2 + dp[i-1-dp[i-1]-1]
				} else {
					dp[i] = dp[i-1] + 2
				}
			}
		}
		if dp[i] > res {
			res = dp[i]
		}
	}
	return res
}